### ISSUE TITLE
Token Sale Verification: TypeError: Cannot read property 'Script' of undefined

### ISSUE THREAD

**Issue by brianlenz** [CityOfZion/neon-wallet#647]
We are doing some final testing of our ICO smart contract on the TestNet, and it looks like we've identified a bug with the "Participate in a token sale" tool. Submissions work just fine when verification is disabled, but they are failing 100% of the time in our tests when verification is enabled.  I'm confident the issue isn't that the actual verification is failing, as we can successfully execute and pass verification by submitting the same transactions in neo-python. The issue seems to be isolated to Neon wallet.

I got a local dev copy of neon-wallet running and did some debugging to try to figure out what is going on and tracked it to this line:

https://github.com/CityOfZion/neon-wallet/blob/master/app/modules/sale.js#L112

I was able to identify the error by including the `error.stack` in the displayed message. Here's the beginning of the (not-so-helpful & incomplete) stack:

```
TypeError: Cannot read property 'Script' of undefined
    at A (http://localhost:3000/dist/bundle.js:76367:130484)
    at http://localhost:3000/dist/bundle.js:76367:130125
    at ...
```

If you want to try it out, here's the script hash on TestNet that is reproducing the issue: ae01c7a5323a4ff6c721fbde741dbd88e844c147

I have very little experience with react, or else I probably could have provided more useful details.

We're hoping to get this issue resolved as soon as possible, so please let me know if there's anything I can do to help!

**Comment by dvdschwrtz** [comment]
Support for verified ICOs is not 100% implemented in Neon wallet. We pushed a quick release for Zeepin and some other ICOs using the non-verification methods. I can get started working on finalizing support for the verified on my long flight tomorrow. I will be on vacation til the 11th but hopefully some other CoZ members can help pull everything together as well. I do not have a ledger with me unfortunately, although I am hoping that I will be able to get one for testing the code I write on the plane during my vacation.

**Comment by brianlenz** [comment]
Thanks for the response, @dvdschwrtz! What's the general scope of remaining work for verification? Is the work described in an issue anywhere?

I may be able to try to help (though there will be a learning curve for me to get up to speed with how everything is working). If you have relevant commits to link to and notes/todos about what's left before you head off on vacation, I can take a crack at a PR...

Thanks!

**Comment by brianlenz** [comment]
Oh, by the way, I have a Ledger I can test with, too.

**Comment by dvdschwrtz** [comment]
Ok great, thanks. I will see how far I can get on my flight and then send you an update :)

**Comment by dvdschwrtz** [comment]
There is no issues or todos but I will try and be helpful with detailed information if there is more work to do

**Comment by mhuggins** [comment]
@brianlenz What is the date of your ICO that you're hoping for this to work?  There are some significant changes in the `dev` branch that we work on, and some other changes need to be addressed beyond just fixing this bug before we can safely release, so I want to make sure I can help coordinate around those changes.

**Comment by brianlenz** [comment]
@mhuggins, we are planning to deploy the SC on Monday and start collecting our pre-sale contributions. The public sale opens Feb. 13. Thanks for your support!

**Comment by dvdschwrtz** [comment]
@brianlenz I am hoping its an easy fix. can you please add this address to the list of verified senders? ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW

**Comment by dvdschwrtz** [comment]
PS: It was nice meeting you this past week. I am hoping that as narrative grows, they can become part of the CoZ community as well. Would be great to have your team involved in CoZ repos and activities :)

**Comment by dvdschwrtz** [comment]
the .Script of undefined was an import error with the TxAttrUsage variable in neon-js. We have not needed it til now so have not caught this minor issue. I am hoping it works with this change but need to be verified to test

**Comment by dvdschwrtz** [comment]
@brianlenz please let me know by pinging me through "@dvdschwrtz" when you have whitelisted my testing address so I can verify it is working for your token. Thanks.

**Comment by brianlenz** [comment]
Thanks @dvdschwrtz! The sale ended on that SC, so I'm deploying a new SC for you to test with now. Will post back once it's up and you're whitelisted...

**Comment by brianlenz** [comment]
OK, @dvdschwrtz, you should be able to test with our contract now here: `e7f761033ff4b2b39ba487a3b102b290ff3e36c2`. Let me know how it goes 👍 Thanks!

**Comment by dvdschwrtz** [comment]
I fixed that issue and everything is looking good on my side after that fix. I am unable to get the verification to pass right now though. you can try pointing to this neon-js in your neon-wallet and test it as well. "neon-js":
      "git+https://github.com/dvdschwrtz/neon-js.git#0c8de6d727ca82e027809f5cb194accd72655136",

I can ask localhuman to take a quick look at this over the next few days potentially. he might have more insights. can you tell me which branch/ repo you are working off. I have looked through the NEX template and can take a look at yours. Did you make any major changes to the contract?

**Comment by dvdschwrtz** [comment]
@brianlenz I have confirmed that the latest package.json changes in the link above work correctly with the NEX template that is using verification. So I would like to review your smart contract

**Comment by dvdschwrtz** [comment]
btw, as a user, if I send NEO to a smart contract that requires verification using the non-verification radio button it sends the NEO even if I am not whitelisted. Using the radio button for verification will block this from sending anything if the user is not whitelisted. we are going to move away from all non-verified options soon but for now this is a critical distinction that you need to clarify with your investors. Of course, we also need to get it working first for this to matter

**Comment by dvdschwrtz** [comment]
I am going to discord. bjornborg

**Comment by brianlenz** [comment]
Sure thing! Project is here:

https://github.com/NarrativeNetwork/tokensale-neo-smartcontract

The script hash I sent you is running the `master` branch, with the only difference being the `original_owner` script hash.

The overall file structure of the contract is similar to the NEX template, but we've added new rules about contract phases, new functionality about minting of tokens, and new utilities for doing things like changing the owner, de-registering KYC addresses, pausing/resuming the sale, and a few others.

Yep, fully understood on the distinction between verified and non-verified contributions. This is exactly why we want to have verification working for our sale. I completely agree with the idea of always requiring verification going forward!

**Comment by anfn101** [comment]
See #658

**Comment by brianlenz** [comment]
FWIW, this is technically different than #658, which is Ledger-specific. This particular issue affected all verified contributions, including non-Ledger wallets. @dvdschwrtz did apply a fix to this through neon-js, though, so it should be fixed in the next Neon wallet release nonetheless.

**Comment by mhuggins** [comment]
@anfn101 @brianlenz I'm going to reopen this one in particular.  Brian's been active in Slack/Discord and has a strong grasp on the issues around this.

**Comment by brianlenz** [comment]
Oh, sorry if I confused things. I do believe that this issue is resolved on the latest `master` after upgrading to neon-js 2.3.4 (courtesy of @dvdschwrtz and @snowypowers). I was just clarifying that this particular issue was a different problem (and thus had a different fix) than #658, which was Ledger-specific.

So to be completely clear: I believe that both this issue and #658 are resolved on the latest master 😁

Sorry if I muddied the waters in my reply...

**Comment by anfn101** [comment]
Will close based on Brian's comment above...
